### PR TITLE
TGP-189 - Create Supplementary Unit Page

### PR DESCRIPTION
**Pull request by yohan-oc**

### CODE REVIEW

File: app/controllers/SupplementaryUnitController.scala [hunk 1]
```diff
@@ -53,15 +54,31 @@ class SupplementaryUnitController @Inject() (
         case Some(value) => form.fill(value)
       }
 
-      Ok(view(preparedForm, mode, recordId))
+      request.userAnswers.get(RecordCategorisationsQuery) match {
+        case None        => Redirect(routes.JourneyRecoveryController.onPageLoad().url)
+        case Some(query) =>
+          query.records(recordId).measurementUnit match {
```
Comment (ChrisCookOC) [inline app/controllers/SupplementaryUnitController.scala]: not sure we can/should be assuming
query.records(recordId)
is defined?

Could be worth switching to a for/yield like in the other places we get the details out of the recordCategorisationsQuery

File: app/controllers/SupplementaryUnitController.scala [hunk 2]
```diff
@@ -53,15 +54,31 @@ class SupplementaryUnitController @Inject() (
...
+            case None                  => Ok(view(preparedForm, mode, recordId, ""))
```
Comment (ChrisCookOC) [inline app/controllers/SupplementaryUnitController.scala]: Do we want to load the page if there is no measurementUnit?

File: test-utils/generators/SupplementaryUnitAmountGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  def formatDoubleWithCommas(value: Double): String = {
+    val formatted   = f"$value%.6f" // Format with 6 decimal places
+    val parts       = formatted.split("\\.")
+    val integerPart = parts(0).reverse.grouped(3).mkString(",").reverse // Add commas every three digits
+    val decimalPart = if (parts.length > 1) "." + parts(1) else "" // Append decimal part if present
+    integerPart + decimalPart
+  }
+
+  def genDoublesperseString(gen: Gen[String], value: String, frequencyV: Int = 1, frequencyN: Int = 10): Gen[String] = {
```
Comment (ChrisCookOC) [inline test-utils/generators/SupplementaryUnitAmountGenerator.scala]: what does Doublesperse mean?

File: app/views/SupplementaryUnitView.scala.html [hunk 4]
```diff
@@ -37,10 +39,11 @@
         @govukInput(
             InputViewModel(
                 field = form("value"),
-                label = LabelViewModel(messages("supplementaryUnit.heading")).asPageHeading()
+                label = Label(content = "supplementaryUnit.heading", classes = "govuk-visually-hidden")
             )
             .asNumeric()
-            .withWidth(Fixed10)
+            .withWidth(Fixed5)
```
Comment (ChrisCookOC) [inline app/views/SupplementaryUnitView.scala.html]: This isn't big enough to support the -9999999999.999999 we support?

File: app/controllers/SupplementaryUnitController.scala [hunk 5]
```diff
@@ -53,15 +54,31 @@ class SupplementaryUnitController @Inject() (
...
+      }
   }
 
   def onSubmit(mode: Mode, recordId: String): Action[AnyContent] =
     (identify andThen getData andThen requireData).async { implicit request =>
       form
         .bindFromRequest()
         .fold(
-          formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode, recordId))),
+          formWithErrors =>
```
Comment (ChrisCookOC) [inline app/controllers/SupplementaryUnitController.scala]: Some formatting stuff happens with big numbers

Enter 1234567890.123456
image

It turns it into scientific notation
image

Click Change

it is still in scientific notation
image

Without changing the value, if I click continue its unhappy with the format
image

Should probably ensure its displayed in full form everytime?

File: test-utils/generators/SupplementaryUnitAmountGenerator.scala (hunk 3 above)
Comment (yohan-oc) [reply test-utils/generators/SupplementaryUnitAmountGenerator.scala]: Name supposed to be gen**Intersperse**String. This is now unused method. Removing this :)

Comment (ChrisCookOC) [reply test-utils/generators/SupplementaryUnitAmountGenerator.scala]: Oh haha. Find and Replace strikes again 😆

File: app/views/SupplementaryUnitView.scala.html (hunk 4 above)
Comment (yohan-oc) [reply app/views/SupplementaryUnitView.scala.html]: This is according to Prototype. https://trader-goods-profile-prototype-18527416d281.herokuapp.com/version-4/categorisation-longer-code-supp-units/supplementary-unit-longer-supp
 
 I will raise this to design team

File: test-utils/generators/SupplementaryUnitAmountGenerator.scala (hunk 3 above)
Comment (yohan-oc) [reply test-utils/generators/SupplementaryUnitAmountGenerator.scala]: facepalm!

File: app/views/SupplementaryUnitView.scala.html (hunk 4 above)
Comment (yohan-oc) [reply app/views/SupplementaryUnitView.scala.html]: Design team agreed to increase the textbox length. Changed back to `.withWidth(Fixed10)`

File: app/controllers/SupplementaryUnitController.scala (hunk 5 above)
Comment (yohan-oc) [reply app/controllers/SupplementaryUnitController.scala]: This is now fixed. I have changed formProvider from double to String

File: app/controllers/SupplementaryUnitController.scala (hunk 1 above)
Comment (yohan-oc) [reply app/controllers/SupplementaryUnitController.scala]: Updated to for/yield

File: app/controllers/SupplementaryUnitController.scala (hunk 2 above)
Comment (yohan-oc) [reply app/controllers/SupplementaryUnitController.scala]: Yes. measurementUnit can be null https://api.trade-tariff.service.gov.uk/green-lanes.html#get-green-lanes-goods-nomenclatures-id

**Review by ChrisCookOC — DISMISSED** [review]

**Review by ChrisCookOC — APPROVED** [review]

**Review by vinoocs — APPROVED** [review]

**Pull request merged**